### DISCUSSION TITLE
Will WindiCSS infer variants created from a plugin via addVariant?

### DISCUSSION THREAD

**Discussion by ahku** [windicss/windicss#171] [General]
Hi, first time posting here, usually I just throw questions out on Twitter but figured I should be more active here instead…

Anyway, I'm trying to migrate a Tailwind project where I have custom variants created via plugin. I've skimmed through the documentation for solutions but could not find anything covering variants.

Here's the code:

``` js
// tailwind.config.js
module.exports = {
  // ...
  plugins: [
      plugin(function ({ addVariant }) {
        addVariant('pointer-hover', ({ modifySelectors, separator }) => {
          modifySelectors(({ className }) => {
            return `.no-touch .${e(
              `pointer-hover${separator}${className}`,
            )}:hover:not(:focus)`
          })
        })
        // ...
    }),
  ]
}
```

The above plugin used to output css such as:

```css
.no-touch .pointer-hover\:bg-gray-200:hover:not(:focus) { ... }
```

Is this supported at all?

**Comment by voorjaar** [comment]
In fact, our interface is more concise.

```js
  plugins: [
      plugin(function ({ addVariant }) {
        addVariant('pointer-hover', ({ modifySelectors }) => {
          return modifySelectors(({ className }) => {
            return `.no-touch .${className}:hover:not(:focus)`;
          });
        });
        // ...
      }),
    ],
```

**Reply by voorjaar** [reply]
className represent `pointer-hover:bg-gray-200`, and we will auto escape it.

**Reply by ahku** [reply]
Wow! That's a much nicer syntax!

I only had an issue with this one:

```js
addVariant('pointer-group-hover', ({ modifySelectors }) => {
  modifySelectors(({ className }) => {
    return `.no-touch .group:hover .${className}`
  })
})
```

`.pointer-group-hover:bg-gray-200` should output:

`.no-touch .group\:hover .pointer-group-hover\:bg-gray-200 {}`

Instead I get:

`.pointer-group-hover\:bg-gray-200`

Maybe that's a bug?

**Reply by voorjaar** [reply]
One mistake. You should always write a return before modifySelectors, because it actually return a style, not just simply adding it. So you should return that function

```js
addVariant('pointer-group-hover', ({ modifySelectors }) => {
  return modifySelectors(({ className }) => {
    return `.no-touch .group:hover .${className}`
  })
})
```

**Reply by ahku** [reply]
I got blinded. Now everything works as expected. Big thanks.
